jwst/associations/lib/rules_level2b.py

```diff
@@ -599,195 +600,42 @@ class Asn_Lv2WFSS(
     def __init__(self, *args, **kwargs):
 
         self.constraints = Constraint([
-
-            # Basic constraints
             Constraint_Base(),
             Constraint_Target(),
-            DMSAttrConstraint(
-                name='instrument',
-                sources=['instrume']
-            ),
-
-            # Allow WFSS exposures but account for the direct imaging.
             Constraint([
-
-                # Constrain on the WFSS exposure
-                Constraint([
-                    DMSAttrConstraint(
-                        name='exp_type',
-                        sources=['exp_type'],
-                        value='nis_wfss|nrc_wfss',
-                    ),
-                    Constraint_Mode(),
-                    Constraint_Single_Science(self.has_science),
-                ]),
-
-                # Or select related imaging exposures.
+                DMSAttrConstraint(
+                    name='exp_type',
+                    sources=['exp_type'],
+                    value='nrc_wfss|nis_wfss',
+                ),
                 DMSAttrConstraint(
                     name='image_exp_type',
                     sources=['exp_type'],
-                    value='nis_image|nrc_image',
-                    force_reprocess=ProcessList.EXISTING,
+                    value='nrc_image|nis_image',
+                    force_reprocess=ProcessList.NONSCIENCE,
                     only_on_match=True,
                 ),
+            ], reduce=Constraint.any),
+            DMSAttrConstraint(
+                name='instrument',
+                sources=['instrume'],
+            ),
+            DMSAttrConstraint(
+                name='mode',
+                sources=['channel', 'pupil'],
```

Comment:
For NIRCam, the `channel` keyword designates the wavelength channel of the instrument (short vs. long). The grisms only exist in the long wavelength channel and hence I would assume that the direct images that go with them would also normally be from the long channel, but I'm not sure that's strictly enforced anywhere (maybe it is in APT?). But I assume the purpose of this constraint is to just remove NIRCam short-wave exposures from consideration when building the level-2b ASN, right?

Reply:
Yes, and in fact, in the latest push, this is changed to "detector" to make it even more restrictive. Speeds up association generation by about 40%. As best as I can tell with NIRCam, one always gets both short and long with any exposures, so one will always have a long channel, and in fact, the same long detector, as the science.

This was the problem originally; such constraints were not being made in the Constraints part of the rules, but after the fact. As a result, all short channel exposures were being added to the mix. This just made an aweful lot of extra work, with a 5X factor occurring with each actual grism exposure. Hence the exponential of the runtime.